server/src/test/java/org/cloudfoundry/identity/uaa/oauth/RefreshRotationTest.java

```diff
@@ -164,6 +198,35 @@ void testRefreshPublicClientWithoutRotation() {
     assertEquals("Refresh without client authentication not allowed.", exception.getMessage());
   }
 
+  @Test
+  @DisplayName("Refresh with allowpublic and rotation but existing token was not public")
+  void testRefreshPublicClientButExistingTokenWasEmptyAuthentication() {
+    BaseClientDetails clientDetails = new BaseClientDetails(tokenSupport.defaultClient);
+    clientDetails.setAutoApproveScopes(singleton("true"));
+    tokenSupport.clientDetailsService.setClientDetailsStore(IdentityZoneHolder.get().getId(), Collections.singletonMap(CLIENT_ID, clientDetails));
+    AuthorizationRequest authorizationRequest = new AuthorizationRequest(CLIENT_ID, tokenSupport.requestedAuthScopes);
+    authorizationRequest.setResourceIds(new HashSet<>(tokenSupport.resourceIds));
+    Map<String, String> azParameters = new HashMap<>(authorizationRequest.getRequestParameters());
+    azParameters.put(GRANT_TYPE, GRANT_TYPE_AUTHORIZATION_CODE);
+    authorizationRequest.setRequestParameters(azParameters);
+    authorizationRequest.setExtensions(Map.of(CLIENT_AUTH_METHOD, CLIENT_AUTH_EMPTY));
+    OAuth2Request oAuth2Request = authorizationRequest.createOAuth2Request();
+    OAuth2Authentication authentication = new OAuth2Authentication(oAuth2Request, tokenSupport.defaultUserAuthentication);
+    new IdentityZoneManagerImpl().getCurrentIdentityZone().getConfig().getTokenPolicy().setRefreshTokenRotate(true);
```

Comment:
The default value (aka if without this line) of `tokenPolicy.refreshTokenRotate` is `false` based on my debugging, so this line does change something. However, commenting out this line, the test still pass.

Reply:
yes because it was a new feature from https://github.com/cloudfoundry/uaa/issues/1448 . 
Clients need to follow the exchange on their side as well. If a client stores a refresh token only but does not care of a new value when using calling grant_type=refresh_token therefore a new feature and as always new feature are off by default.